dm/dm-best-practices.md

```diff
@@ -0,0 +1,234 @@
+
+- Non-clustered indexes + `shard row id bit`
+
+    Using non-clustered indexes and `shard row id bit`, you can avoid the write hotspot problem when using `AUTO_INCREMENT`. However, table lookup in this scenario can impact the query performance when querying using the primary key.
+
+- Clustered indexes + external distributed ID generators
+
+    If you want to use clustered indexes and keep the IDs consecutive, consider using external distributed ID generators, such as the Snowflake algorithm and Leaf. The application program generates sequence IDs, which can guarantee that the IDs are consecutive to a certain extent. It also retains the benefits of using clustered indexes. But you need to customize the applications.
+
+- Clustered indexes + `AUTO_RANDOM`
+
+    This solution can retain the benefits of using clustered indexes and avoid the write hotspot problem. It requires less effort for customization. You can modify the schema attribute when you switch to use TiDB as the write database. In subsequent queries, if you have to use the ID column to sort data, you can use the [`AUTO_RANDOM`](/auto-random.md) ID column and left shift 5 bits to ensure the order of the query data. For example:
+
+    ```sql
+    CREATE TABLE t (a bigint PRIMARY KEY AUTO_RANDOM, b varchar(255));
+    Select  a, a<<5 ,b from t order by a <<5 desc
+    ```
+
+The following table summarizes the pros and cons of each solution.
+
+| Scenario | Recommended solution | Pros | Cons |
+| :--- | :--- | :--- | :--- |
+| <li>TiDB will act as the primary and write-intensive database. </li><li>The business logic strongly depends on the continuity of the primary key IDs.</li> | Create the table with non-clustered indexes and set `SHARD_ROW_ID_BIT`. Use `SEQUENCE` as the primary key column.  | It can avoid data write hotspots and ensure the continuity and monotonic increment of business data. | <li>The throughput capacity of data write is decreased to ensure data write continuity. </li><li>The performance of primary key queries is decreased.</li> |
+| <li>TiDB will act as the primary and write-intensive database. </li><li>The business logic strongly depends on the increment of the primary key IDs.</li>  | Create the table with non-clustered indexes and set `SHARD_ROW_ID_BIT`. Use an application ID generator to generate the primary key IDs. | It can avoid data write hotspots, guarantee the performance of data write, guarantee the increment of business data, but cannot guarantee continuity. | <li>You need to customize the application. </li><li>External ID generators strongly depend on the clock accuracy and might introduce failure risks.</li> |
```

Comment:
```suggestion
| <li>TiDB will act as the primary and write-intensive database. </li><li>The business logic strongly relies on the increment of the primary key IDs.</li>  | Create tables with non-clustered indexes and set `SHARD_ROW_ID_BIT`. Use an application ID generator to generate the primary key IDs. | It can avoid data write hotspots, guarantee the performance of data write, and guarantee the increment of business data, but cannot guarantee continuity. | <li>You need to customize the application. </li><li>External ID generators strongly rely on the clock accuracy and might introduce failures.</li> |
```